Sources/MapboxMaps/Location/Puck/Puck2D.swift

```diff
@@ -0,0 +1,169 @@
+            }
+            layer.location = .constant([
+                location.coordinate.latitude,
+                location.coordinate.longitude,
+                location.location.altitude
+            ])
+            layer.locationTransition = StyleTransition(duration: 0.5, delay: 0)
+            layer.topImageSize = configuration.resolvedScale
+            layer.bearingImageSize = configuration.resolvedScale
+            layer.shadowImageSize = configuration.resolvedScale
+            layer.emphasisCircleRadiusTransition = StyleTransition(duration: 0, delay: 0)
+            layer.bearingTransition = StyleTransition(duration: 0, delay: 0)
+            if configuration.showsAccuracyRing {
+                layer.accuracyRadius = .constant(location.horizontalAccuracy)
+                layer.accuracyRadiusColor = .constant(StyleColor(UIColor(red: 0.537, green: 0.812, blue: 0.941, alpha: 0.3)))
+                layer.accuracyRadiusBorderColor = .constant(StyleColor(.lightGray))
+            }
+            switch puckBearingSource {
+            case .heading:
+                layer.bearing = .constant(location.headingDirection ?? 0)
+            case .course:
+                layer.bearing = .constant(location.course)
+            }
+        case .reducedAccuracy:
+            fallthrough
+        @unknown default:
+            layer.accuracyRadius = .expression(Exp(.interpolate) {
+                Exp(.linear)
+                Exp(.zoom)
+                0
+                400000
+                4
+                200000
+                8
+                5000
+            })
+            layer.accuracyRadiusColor = .constant(StyleColor(UIColor(red: 0.537, green: 0.812, blue: 0.941, alpha: 0.3)))
+            layer.accuracyRadiusBorderColor = .constant(StyleColor(.lightGray))
+        }
+
+        // Create the properties dictionary for the updated layer
+        let newLayerProperties = try! layer.jsonObject()
+        // Construct the properties dictionary to reset any properties that are no longer used
+        let unusedPropertyKeys = previouslySetLayerPropertyKeys.subtracting(newLayerProperties.keys)
+        let unusedProperties = Dictionary(uniqueKeysWithValues: unusedPropertyKeys.map { (key) -> (String, Any) in
+            (key, Style.layerPropertyDefaultValue(for: .locationIndicator, property: key).value)
+        })
+        // Merge the new and unused properties
+        let allLayerProperties = newLayerProperties.merging(unusedProperties, uniquingKeysWith: { $1 })
+        // Store the new set of property keys
+        previouslySetLayerPropertyKeys = Set(newLayerProperties.keys)
+
+        // Update or add the layer
+        if style.layerExists(withId: Self.layerID) {
+            try! style.setLayerProperties(for: Self.layerID, properties: allLayerProperties)
+        } else {
+            try! style.addPersistentLayer(with: allLayerProperties, layerPosition: nil)
+        }
```

Comment:
Can we extract that code and provide a comment over the reason to make that dance?

Reply:
What would you put in the comment? I was thinking that the existing in-line comments would be sufficient.

As for extraction, I created a ticket to create an abstraction around this that can be reused here, in annotation managers, and in Style.updateLayer. I wouldn't want to put this in a separate method just to use it from this one place — that'd actually be detrimental to readability because if you were reading or refactoring that method you'd then have to do a search to see how many different places it was used. As-is, it's immediately clear from the context that there's only one place where this code is used.

Reply:
During the call, you mentioned the reason to make sets and calculate the difference. It was very useful to understand so it would be cool to see it as a comment.

Reply:
Added a comment: https://github.com/mapbox/mapbox-maps-ios/pull/765/commits/da7ea423269c9cb00dd214bfad6b895483d7abbe

Look good?

Reply:
Perfect! Thank you.